src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+  using namespace aten;
+  const IdxType* indptr = static_cast<IdxType*>(mat.indptr->data);
+  const IdxType* indices = static_cast<IdxType*>(mat.indices->data);
+  const IdxType* data = CSRHasData(mat)? static_cast<IdxType*>(mat.data->data) : nullptr;
+  const IdxType* rows_data = static_cast<IdxType*>(rows->data);
+  const IdxType* nids = NIDs->shape[0] ? static_cast<IdxType*>(NIDs->data) : nullptr;
+  const auto num_rows = rows->shape[0];
+  const IdxType num_picks = __num_picks;
+  const auto& ctx = mat.indptr->ctx;
+
+  constexpr bool linear_c_convergence = false;
+
+  const bool weights = !IsNullArray(prob);
+  if (importance_sampling >= 0)
+    importance_sampling += weights;
+  auto A_arr = prob;
+  FloatType *A = static_cast<FloatType*>(A_arr->data);
+  constexpr double eps = 0.0001;
+  constexpr FloatType ONE = 1;
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  FloatArray cs_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* cs = static_cast<FloatType*>(cs_array->data);
+  FloatArray ds_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* ds = static_cast<FloatType*>(ds_array->data);
+
+  IdxType max_d = 1;
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_d = indptr[rid + 1] - indptr[rid];
+    max_d = std::max(act_d, max_d);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_d;
+    cs[i] = num_picks / d;
+    ds[i] = d;
+    hop_size += act_d;
+  }
+
+  if (importance_sampling) {
+    FloatArray ps_array = NDArray::Empty({max_d + 1},
+                                          DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                          ctx);
+    FloatType* ps = static_cast<FloatType*>(ps_array->data);
+
+    double prev_ex_nodes = max_d * num_rows;
+
+    phmap::flat_hash_map<IdxType, FloatType> hop_map2;
+    for (int iters = 0; iters < importance_sampling || importance_sampling < 0; iters++) {
+      if (!weights || iters) {
+        hop_map2.clear();
+        for (int64_t i = 0; i < num_rows; ++i) {
+          const FloatType c = cs[i];
+          const IdxType rid = rows_data[i];
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+            const auto ct = c * (weights && iters == 1 ? A[j] : 1);
+            auto itb = hop_map2.emplace(indices[j], ct);
+            if (!itb.second)
+              itb.first->second = std::max(ct, itb.first->second);
+          }
+        }
+        if (hop_map.empty())
+          hop_map = std::move(hop_map2);
+        else
+          for (auto it : hop_map2)
+            hop_map[it.first] *= it.second;
+      }
+
+      for (int64_t i = 0; i < num_rows; ++i) {
```

Comment:
Can we parallelize this loop?  If so, you can use `runtime::parallel_for`.

Reply:
If a parallel implementation is important, then I can implement the hashtable parts in other ways to make it parallel also. I could do that later also if needed.

Reply:
Also, if the cpu implementation needs to be improved, then one needs to implement a vectorized PRNG so that same logic in the cuda implementation can be used without the use of a hash_map. PCG was slow so I had to reduce the number of calls to it by using a hash_map to cache the computed random numbers.

Reply:
> Also, if serial parts aren't parallelized, then it will be a very huge bottleneck, and nobody should use it like that.

So you are suggesting that this for loop is actually computationally lighter than the other parts?  If that's the case, I think you can put it as a comment.

> Maybe I can make a proper parallel CPU implementation in a later PR? I think it will require a huge change to the code.

You don't have to.

> Also, if the cpu implementation needs to be improved, then one needs to implement a faster PRNG with same capability so that same logic in the cuda implementation can be used without the use of a hash_map. PCG was slow so I had to reduce the number of calls to it by using a hash_map to cache the computed random numbers.

I have a separate question on why to use PCG instead of STL's RNG but we can move our discussion there.